cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/LeaderReplicaDistributionGoal.java

```diff
@@ -0,0 +1,343 @@
+                                           optimizedGoals,
+                                           optimizationOptions);
+      // Only check if we successfully moved something.
+      if (b != null) {
+        if (--numLeaderReplicas <= _balanceUpperLimit) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  private boolean rebalanceByMovingLeadershipIn(Broker broker,
+                                                ClusterModel clusterModel,
+                                                Set<Goal> optimizedGoals,
+                                                OptimizationOptions optimizationOptions) {
+    if (optimizationOptions.excludedBrokersForLeadership().contains(broker.id())) {
+      return true;
+    }
+
+    int numLeaderReplicas = broker.leaderReplicas().size();
+    Set<Broker> candidateBrokers =  Collections.singleton(broker);
+    for (Replica replica : broker.replicas()) {
+      if (replica.isLeader()) {
+        continue;
+      }
+      Broker b = maybeApplyBalancingAction(clusterModel,
+                                           clusterModel.partition(replica.topicPartition()).leader(),
+                                           candidateBrokers,
+                                           ActionType.LEADERSHIP_MOVEMENT,
+                                           optimizedGoals,
+                                           optimizationOptions);
+      // Only check if we successfully moved something.
+      if (b != null) {
+        if (++numLeaderReplicas >= _balanceLowerLimit) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  private boolean rebalanceByMovingReplicasOut(Broker broker,
+                                               ClusterModel clusterModel,
+                                               Set<Goal> optimizedGoals,
+                                               OptimizationOptions optimizationOptions) {
+    // Get the eligible brokers.
+    SortedSet<Broker> candidateBrokers;
+    if (_selfHealingDeadBrokersOnly) {
+      candidateBrokers = new TreeSet<>(Comparator.comparingInt((Broker b) -> b.replicas().size())
+                                                 .thenComparingInt(Broker::id));
+      candidateBrokers.addAll(clusterModel.aliveBrokers());
+    } else {
+      candidateBrokers = new TreeSet<>(Comparator.comparingInt((Broker b) -> b.leaderReplicas().size())
+                                                 .thenComparingInt(Broker::id));
+      candidateBrokers.addAll(clusterModel.aliveBrokers()
+                      .stream()
+                      .filter(b -> b.leaderReplicas().size() < _balanceUpperLimit)
+                      .collect(Collectors.toSet()));
+    }
+
+    int balanceUpperLimit = _selfHealingDeadBrokersOnly ? 0 : _balanceUpperLimit;
+    Set<String> excludedTopics = optimizationOptions.excludedTopics();
+    Set<Replica> candidateReplicas = _selfHealingDeadBrokersOnly ? new HashSet<>(broker.replicas())
```

Comment:
Nit (`style`): `Set<Replica> candidateReplicas = new HashSet<>(_selfHealingDeadBrokersOnly ? broker.replicas() : broker.leaderReplicas());`

-- It is also perfectly OK to keep it as is, but typically I try to extract common logic when using ternary.